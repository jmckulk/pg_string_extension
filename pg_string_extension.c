#include "postgres.h"
#include <string.h>
#include "fmgr.h"
#include "utils/geo_decls.h"
#include "utils/builtins.h"

#ifdef PG_MODULE_MAGIC
PG_MODULE_MAGIC;
#endif

PG_FUNCTION_INFO_V1(echo_text);

Datum
echo_text(PG_FUNCTION_ARGS){
  text *t = PG_GETARG_TEXT_PP(0);
  text *new_t = (text *) palloc(VARSIZE_ANY_EXHDR(t) + VARHDRSZ);
  SET_VARSIZE(new_t, VARSIZE_ANY_EXHDR(t) + VARHDRSZ);
  memcpy( (void *) VARDATA(new_t),
          (void *) VARDATA_ANY(t),
          VARSIZE_ANY_EXHDR(t));
  PG_RETURN_TEXT_P(new_t);
}


PG_FUNCTION_INFO_V1(is_palindrome);

Datum
is_palindrome(PG_FUNCTION_ARGS){
  text *word = PG_GETARG_TEXT_PP(0);

  char * c_word = text_to_cstring(word);
  int len = strlen(c_word);

  if(len == 0){
    PG_RETURN_BOOL(false);
  }

  if(len == 1){
    PG_RETURN_BOOL(true);
  }

  int front = 0, back = len - 1;

  while((front < back) && (front != back)){
    if(c_word[front] != c_word[back]){
      PG_RETURN_BOOL(false);
    } else {
      front++;
      back--;
    }
  }

  PG_RETURN_BOOL(true);
}
