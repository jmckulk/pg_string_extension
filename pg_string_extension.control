# PostgreSQL String Extension
comment = 'string functions for PostgreSQL'
default_version = '0.1'
relocatable = false
