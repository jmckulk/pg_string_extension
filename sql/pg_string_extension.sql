CREATE EXTENSION pg_string_extension;
SELECT is_palindrome('');
SELECT is_palindrome('a');
SELECT is_palindrome('aa');
SELECT is_palindrome('aba');
SELECT is_palindrome('helloolleh');
SELECT is_palindrome('world');
SELECT is_palindrome('aaab');
SELECT is_palindrome('helloollehhelloollehhelloollehhelloollehhelloollehhelloollehhelloollehhelloollehhelloollehhelloolleh');
