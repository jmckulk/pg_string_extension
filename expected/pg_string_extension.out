CREATE EXTENSION pg_string_extension;
SELECT is_palindrome('');
 is_palindrome 
---------------
 f
(1 row)

SELECT is_palindrome('a');
 is_palindrome 
---------------
 t
(1 row)

SELECT is_palindrome('aa');
 is_palindrome 
---------------
 t
(1 row)

SELECT is_palindrome('aba');
 is_palindrome 
---------------
 t
(1 row)

SELECT is_palindrome('helloolleh');
 is_palindrome 
---------------
 t
(1 row)

SELECT is_palindrome('world');
 is_palindrome 
---------------
 f
(1 row)

SELECT is_palindrome('aaab');
 is_palindrome 
---------------
 f
(1 row)

SELECT is_palindrome('helloollehhelloollehhelloollehhelloollehhelloollehhelloollehhelloollehhelloollehhelloollehhelloolleh');
 is_palindrome 
---------------
 t
(1 row)

